strawberry/tools/depth_limit_validator.py

```diff
@@ -0,0 +1,186 @@
+    definitions: List[DefinitionNode],
+) -> Dict[str, FragmentDefinitionNode]:
+    fragments = {}
+    for definition in definitions:
+        if isinstance(definition, FragmentDefinitionNode):
+            fragments[definition.name.value] = definition
+
+    return fragments
+
+
+# This will actually get both queries and mutations.
+# We can basically treat those the same
+def get_queries_and_mutations(
+    definitions: List[DefinitionNode],
+) -> Dict[str, OperationDefinitionNode]:
+    operations = {}
+
+    for definition in definitions:
+        if isinstance(definition, OperationDefinitionNode):
+            operations[
+                definition.name.value if definition.name else "anonymous"
+            ] = definition
+
+    return operations
+
+
+def determine_depth(
+    node: Node,
+    fragments: Dict[str, FragmentDefinitionNode],
+    depth_so_far: int,
+    max_depth: int,
+    context: ValidationContext,
+    operation_name: str,
+    ignore: Optional[List[IgnoreType]] = None,
+) -> int:
+    if depth_so_far > max_depth:
+        context.report_error(
+            GraphQLError(
+                f"'{operation_name}' exceeds maximum operation depth of {max_depth}",
+                [node],
+            )
+        )
+        return depth_so_far
+
+    if isinstance(node, FieldNode):
+        # by default, ignore the introspection fields which begin with double underscores
+        should_ignore = is_instrospection_key(node.name.value) or see_if_ignored(
+            node, ignore
+        )
+
+        if should_ignore or not node.selection_set:
+            return 0
+
+        return 1 + max(
+            map(
+                lambda selection: determine_depth(
+                    selection,
+                    fragments,
+                    depth_so_far + 1,
+                    max_depth,
+                    context,
+                    operation_name,
+                    ignore,
+                ),
+                node.selection_set.selections,
+            )
+        )
+    elif isinstance(node, FragmentSpreadNode):
+        return determine_depth(
+            fragments[node.name.value],
+            fragments,
+            depth_so_far,
+            max_depth,
+            context,
+            operation_name,
+            ignore,
+        )
+    elif (
+        isinstance(node, InlineFragmentNode)
+        or isinstance(node, FragmentDefinitionNode)
+        or isinstance(node, OperationDefinitionNode)
+    ):
+        return max(
+            map(
+                lambda selection: determine_depth(
+                    selection,
+                    fragments,
+                    depth_so_far,
+                    max_depth,
+                    context,
+                    operation_name,
+                    ignore,
+                ),
+                node.selection_set.selections,
+            )
+        )
+    else:
+        raise Exception(f"uh oh! depth crawler cannot handle: {node.kind}")
```

Comment:
I guess this would be hard to test since this function currently appears to handle all kinds of nodes.

```suggestion
        raise Exception(f"Depth crawler cannot handle: {node.kind}")  # pragma: no cover
```